### ISSUE TITLE
PullerInternal processInbox Issue

### ISSUE THREAD

**Issue by bundeeteddee** [couchbase/couchbase-lite-java-core#707]
Hi there,

I've been trying to find a sure way to replicate this but nothing concrete at all. This is not a crash but something that will block my entire db operation until everything is done. The output is as follows:
```
08-20 14:43:16.834   W/Sync﹕ com.couchbase.lite.replicator.PullerInternal@2e155e00 no new remote revisions to fetch.  add lastInboxSequence (20168) to pendingSequences (com.couchbase.lite.support.SequenceMap@15bd3639)
08-20 14:43:16.834   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 0 0 0 0
08-20 14:43:16.894   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
08-20 14:43:16.904   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 2 0 0 0
08-20 14:43:16.914   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 3 0 0 0
08-20 14:43:16.914   W/Sync﹕ com.couchbase.lite.replicator.PullerInternal@2e155e00 no new remote revisions to fetch.  add lastInboxSequence (20171) to pendingSequences (com.couchbase.lite.support.SequenceMap@15bd3639)
08-20 14:43:16.914   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
08-20 14:43:16.924   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
08-20 14:43:16.934   W/Sync﹕ com.couchbase.lite.replicator.PullerInternal@2e155e00 no new remote revisions to fetch.  add lastInboxSequence (20172) to pendingSequences (com.couchbase.lite.support.SequenceMap@15bd3639)
08-20 14:43:16.934   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 0 0 0 0
08-20 14:43:17.004   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
08-20 14:43:17.024   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 2 0 0 0
08-20 14:43:17.034   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 3 0 0 0
08-20 14:43:17.044   W/Sync﹕ com.couchbase.lite.replicator.PullerInternal@2e155e00 no new remote revisions to fetch.  add lastInboxSequence (20175) to pendingSequences (com.couchbase.lite.support.SequenceMap@15bd3639)
08-20 14:43:17.044   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 0 0 0 0
08-20 14:43:17.054   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
08-20 14:43:17.064   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 2 0 0 0
08-20 14:43:17.074   W/Sync﹕ com.couchbase.lite.replicator.PullerInternal@2e155e00 no new remote revisions to fetch.  add lastInboxSequence (20177) to pendingSequences (com.couchbase.lite.support.SequenceMap@15bd3639)
08-20 14:43:17.084   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 0 0 0 0
08-20 14:43:17.124   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
08-20 14:43:17.134   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 3 0 0 0
08-20 14:43:17.134   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 3 0 0 0
08-20 14:43:17.184   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 4 0 0 0
08-20 14:43:17.234   W/Sync﹕ com.couchbase.lite.replicator.PullerInternal@2e155e00 no new remote revisions to fetch.  add lastInboxSequence (20181) to pendingSequences (com.couchbase.lite.support.SequenceMap@15bd3639)
08-20 14:43:17.234   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
08-20 14:43:17.234   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
08-20 14:43:17.244   W/Sync﹕ com.couchbase.lite.replicator.PullerInternal@2e155e00 no new remote revisions to fetch.  add lastInboxSequence (20182) to pendingSequences (com.couchbase.lite.support.SequenceMap@15bd3639)
08-20 14:43:17.244   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 0 0 0 0
08-20 14:43:17.254   A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :141245 : 1 0 0 0
```

I am seeing a ton of these in logcat (The PullerInternal saying no remote revisions to fetch) and it will keep going until the lastInboxSequence goes close to the number of documents i have in the database. The number "141245" is the number of document in my couch database in the tablet (i have about 240k documents in my remote database). The first number in the pattern (i.e. "4 0 0 0") indicates the getChangesCount() from pull replicator object.

**Question 1**: What exactly is it doing? Trying to find the best checkpoint to continue syncing? 
**Question 2**: Any ideas why this is taking forever to finish? Its constantly hitting the remote database as well (constant data usage) either to get _changes or _revs_diff. Usually takes 45min to finish off.

Cheers
Kien

**Comment by hideki** [comment]
Hi @bundeeteddee,

If my memory is correct, you are using CouchDB for server side? Couchbase Lite gets list of change document from the server. I assume there were many document changes at server side.

Can you verify if CouchDB is latest or relatively latest? And what version of Couchbase Lite are you using, v1.1.0?

Thanks,
Hideki

**Comment by bundeeteddee** [comment]
Hi @hideki ,

I'm currently on couchDb 1.5.0, couches lite version i think is very close to master.

The ton of query to get document changes can happen even when there aren't many or any changes at all from the remote server. It can be the application running over night, and when resuming in the morning, the logs keep coming on changes. I have a few assumptions/guesses so far:

1) I have seen this logs like this before (not sure how frequent). But after this happens, i do see the flurry of logs after
```
08-20 15:29:04.695    8525-9056/com.app I/System.out﹕ Thread-10292(ApacheHTTPLog):SmartBonding Enabling is false, SHIP_BUILD is true, log to file is false, DBG is false
08-20 15:29:04.695    8525-9009/com.app A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :244531 : 0 0 0 0
08-20 15:29:04.705    8525-9059/com.app I/System.out﹕ Thread-10295(ApacheHTTPLog):isSBSettingEnabled false
08-20 15:29:04.705    8525-9059/com.app I/System.out﹕ Thread-10295(ApacheHTTPLog):isShipBuild true
08-20 15:29:04.705    8525-9059/com.app I/System.out﹕ Thread-10295(ApacheHTTPLog):SMARTBONDING_ENABLED is false
08-20 15:29:04.705    8525-9059/com.app I/System.out﹕ Thread-10295(ApacheHTTPLog):SmartBonding Enabling is false, SHIP_BUILD is true, log to file is false, DBG is false
08-20 15:29:04.735    8525-8525/com.app D/com.adapters.db.couch.DatabaseAdapter﹕ Ektorp's onSuccess()
08-20 15:29:04.885    8525-9056/com.app I/System.out﹕ CBLRequestWorker-43323-0 calls detatch()
08-20 15:29:04.885    8525-9059/com.app I/System.out﹕ CBLRequestWorker-4cb03-0 calls detatch()
08-20 15:29:04.885    8525-9063/com.app I/System.out﹕ Thread-10297(ApacheHTTPLog):isSBSettingEnabled false
08-20 15:29:04.885    8525-9063/com.app I/System.out﹕ Thread-10297(ApacheHTTPLog):isShipBuild true
08-20 15:29:04.885    8525-9063/com.app I/System.out﹕ Thread-10297(ApacheHTTPLog):SMARTBONDING_ENABLED is false
08-20 15:29:04.885    8525-9063/com.app I/System.out﹕ Thread-10297(ApacheHTTPLog):SmartBonding Enabling is false, SHIP_BUILD is true, log to file is false, DBG is false
08-20 15:29:04.945    8525-9009/com.app A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :244532 : 1 0 0 0
08-20 15:29:04.945    8525-9063/com.app I/System.out﹕ ChangeTracker-https://---:---@<url>:15984/dev calls detatch()
08-20 15:29:04.945    8525-9009/com.app A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :244532 : 1 0 0 0
08-20 15:29:05.445    8525-9009/com.app W/Sync﹕ com.couchbase.lite.replicator.PullerInternal@29bc63df no new remote revisions to fetch.  add lastInboxSequence (246101) to pendingSequences (com.couchbase.lite.support.SequenceMap@91ea82c)
08-20 15:29:05.445    8525-9009/com.app A/com.adapters.db.couch.DatabaseAdapter﹕ Replication >> :244531 : 0 0 0 0
08-20 15:29:06.455    8525-9059/com.app I/System.out﹕ Thread-10295(ApacheHTTPLog):isSBSettingEnabled false
08-20 15:29:06.455    8525-9059/com.app I/System.out﹕ Thread-10295(ApacheHTTPLog):isShipBuild true
08-20 15:29:06.455    8525-9059/com.app I/System.out﹕ Thread-10295(ApacheHTTPLog):SMARTBONDING_ENABLED is false
08-20 15:29:06.455    8525-9059/com.app I/System.out﹕ Thread-10295(ApacheHTTPLog):SmartBonding Enabling is false, SHIP_BUILD is true, log to file is false, DBG is false
08-20 15:29:06.505    8525-9059/com.app I/System.out﹕ CBLRequestWorker-4cb03-0 calls detatch()
08-20 15:29:06.505    8525-9059/com.app W/Sync﹕ com.couchbase.lite.replicator.ReplicationInternal$7@f10971: Database is null or closed, not calling db.setLastSequence()
```
So it seems like the last sequence or a checkpoint was lost, and it has to recalculate/find out by going through it from the very beginning (?).

2) A conflict for something that is local
```
08-19 18:08:38.584  28369-28702/com.app E/RemoteRequest﹕ Got error status: 409 for https://<url>:15984/dev/_local/d3e7e1a3ebd21574ec2b07e776d1da03b7b44479.  Reason: Conflict
08-19 18:08:38.584  28369-28702/com.app W/Sync﹕ com.couchbase.lite.replicator.ReplicationInternal$7@11b037dc: Unable to save remote checkpoint
```
Can something like this happen when you have more than 1 pull replicator on a database or recreating database object when one already exists?

Cheers
Kien

**Comment by bundeeteddee** [comment]
One more thing (and this make a bit more sense to me),

I got a one-shot replication going, mid way through i quit the application completely. When i re-start the application, i do see the log continuously till the last stop point, and the replication continued on as normal. Will this happen if its continuous replication?

Kien

**Comment by hideki** [comment]
Hi @bundeeteddee,

Thank you for additional information.

Can you test this with using v1.1.0? I'd like to know if I introduced new issue between v1.1.0 and current master.

> I got a one-shot replication going, mid way through i quit the application completely. When i re-start the application, i do see the log continuously till the last stop point, and the replication continued on as normal. Will this happen if its continuous replication?

If checkpoint is saved correctly, the replicator continue from last checkpoint.

**Comment by bundeeteddee** [comment]
Yep, i am using v1.1.0 for couch-life-android library. 

What i did:
- start application, create one-shot replication, let it continue till about 8000 documents out of 30k 
- go to task manager and remove the application
- Restart application, create one-shot replication again

Observation:
- From the info listener attached to the pull replicator, it goes through about 200 changes each hit, and seeing the addLastSequence log, climbing from the beginning to about the 8000 mark, and the replication then continues form that point on.

**Comment by hideki** [comment]
Hi @bundeeteddee,

> about 200 changes each hit

This is a batch size.

> From the info listener attached to the pull replicator, it goes through about 200 changes each hit, and seeing the addLastSequence log, climbing from the beginning to about the 8000 mark, and the replication then continues form that point on.

Please confirm following
After restarting app:
- 0 - 8000 documents: Did you observe to ChangeTracker checked update for 8000 docs, but no replication occurred because documents are already replicated to local (client) database.
- 8001 - last documents: Did you observe to replicate documents.

**Comment by bundeeteddee** [comment]
Hi @hideki 

>0 - 8000 documents: Did you observe to ChangeTracker checked update for 8000 docs, but no replication occurred because documents are already replicated to local (client) database.
Yes, while going from 0-8000 documents after restart, from looking at the output coming from the listener attached to the pull replicator object, **getChangesCount()** spits roughly 200 every time and **getCompletedChangesCount()** is 0. And in-between every few output form the listener, there are info logs indicating that it kept adding lastInboxSequence, jumping few hundred at a time from 0 to 8000.

>8001 - last documents: Did you observe to replicate documents.
Yes, replication was back to normal. **getChangesCount()** and **getCompletedChangesCount()** increases together.

Cheers
Kien

**Comment by hideki** [comment]
Hi @borrrden,

I believe CBL client should skip 0-8000 docs check if local checkpoint is saved correctly. I need to re-read codes.

Is it possible to test this with latest version of CouchDB 1.6.1?

Thanks,
Hideki

**Comment by hideki** [comment]
Hi @bundeeteddee,

I talked about this with team members. 

With current implementation, checkpoint is saved after all replication finished with one-shot replication.
So, if replication is killed in middle of replication like your scenario, checkpoint is not saved.
As result, after restarting app, CBL needs to recheck if documents are existed for previously replicated documents. But this process is not too slow because documents are replicated already.

In short period of time, we don't have plan to change how to save checkpoint for one-short replication.

Current log messages makes confusion. We need to improve logging message.

Thanks,
Hideki

**Comment by bundeeteddee** [comment]
Hi @hideki ,

Thanks for the explanation and its makes sense from the observation at the moment (speed wise for 8k documents maybe 2mins or so). May i ask if there are other instances that checkpoints might disappear/overwritten/reset that would result in CBL needing to check from the very beginning?

Kien

**Comment by hideki** [comment]
Hi @bundeeteddee,

I feel 2min/8k docs is slow. If my understanding is correct, CouchDB returns entire document change list by one response, and CBL client parses the response, and process them. I guess this process use a lot of memory for processing 8K docs, and GC takes time to release memory on Android. We have recognized this process should be improved.

> May i ask if there are other instances that checkpoints might disappear/overwritten/reset that would result in CBL needing to check from the very beginning?

checkpoints are used by both pull and push replicators.

**Comment by hideki** [comment]
About log message, message is logged at
https://github.com/couchbase/couchbase-lite-java-core/blob/b9b6fffa3c0b2e702769565eb64227561ab2dc29/src/main/java/com/couchbase/lite/replicator/PullerInternal.java#L197 . This message should be logged with debug or info mode, not with warning mode.

**Comment by hideki** [comment]
Replicate 30K docs from CouchDB

Notes
- CouchDB is v1.6.1 which installed on my OSX
- Android Device is Nexus 6 with API 23 that is in same wifi network with OSX that hosts CouchDB
- Each document is small that only contains one element with coudhbase iite used fields.

To replicate 30K docs from CouchDB 
- 3 min 44 sec

To replicate 30K (remainings 19801 docs) docs with 10199 docs already exists in local without proper check point (scenario of this issue)
- 2min 44 sec

As we expected, Pull replication skips to replicate already existed documents in local db.
